File: ArchivistSDK/src/main/java/com/zorroa/archivist/sdk/util/Json.java
package com.zorroa.archivist.sdk.util;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import com.zorroa.archivist.sdk.exception.MalformedDataException;

import java.io.IOException;

/**
 * Some basic JSON utilities.
 *
 * @author chambers
 *
 */
public class Json {

    public final static ObjectMapper Mapper = new ObjectMapper();
    static {
        Mapper.setSerializationInclusion(JsonInclude.Include.NON_NULL);
        Mapper.configure(SerializationFeature.WRITE_NULL_MAP_VALUES, false);
        Mapper.configure(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS, false);
        Mapper.configure(SerializationFeature.WRITE_ENUMS_USING_INDEX, true);
        Mapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
    }

    public static String serializeToString(Object object) {
        try {
            return Json.Mapper.writeValueAsString(object);
        } catch (JsonProcessingException e) {
            throw new MalformedDataException(
                    "Failed to serialize object, unexpected: " + e, e);
        }
    }

    public static byte[] serialize(Object object) {
        try {
            return Json.Mapper.writeValueAsBytes(object);
        } catch (JsonProcessingException e) {
            throw new MalformedDataException(
                    "Failed to serialize object, unexpected: " + e, e);
        }
    }

    public static String serializeToString(Object object, String onNull) {
        if (object == null) {
            return onNull;
        }
        try {
            return Json.Mapper.writeValueAsString(object);
        } catch (JsonProcessingException e) {
            throw new MalformedDataException(
                    "Failed to serialize object, unexpected: " + e, e);
        }
    }

    public static byte[] serialize(Object object, String onNull) {
        if (object == null) {
            return onNull.getBytes();
        }
        try {
            return Json.Mapper.writeValueAsBytes(object);
        } catch (JsonProcessingException e) {
            throw new MalformedDataException(
                    "Failed to serialize object, unexpected: " + e, e);
        }
    }

    public static <T> T deserialize(byte[] data, Class<T> valueType) {
        try {
            return Json.Mapper.readValue(data, valueType);
        } catch (IOException e) {
            throw new MalformedDataException(
                    "Failed to unserialize object, unexpected " + e, e);
        }
    }

    public static <T> T deserialize(byte[] data, TypeReference<T> valueType) {
        try {
            return Json.Mapper.readValue(data, valueType);
        } catch (IOException e) {
            throw new MalformedDataException(
                    "Failed to unserialize object, unexpected " + e, e);
        }
    }

    public static <T> T deserialize(String data, Class<T> valueType) {
        try {
            return Json.Mapper.readValue(data, valueType);
        } catch (IOException e) {
            throw new MalformedDataException(
                    "Failed to unserialize object, unexpected " + e, e);
        }
    }

    public static <T> T deserialize(String data, TypeReference<T> valueType) {
        try {
            return Json.Mapper.readValue(data, valueType);
        } catch (IOException e) {
            throw new MalformedDataException(
                    "Failed to unserialize object, unexpected " + e, e);
        }
    }
}

File: services/auth-server/auth-client/src/main/kotlin/com/zorroa/auth/client/Json.kt
package com.zorroa.auth.client

import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.core.JsonParser
import com.fasterxml.jackson.databind.DeserializationFeature
import com.fasterxml.jackson.databind.MapperFeature
import com.fasterxml.jackson.databind.SerializationFeature
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import java.text.SimpleDateFormat
import okhttp3.MediaType

object Json {

    val mapper = jacksonObjectMapper()

    init {
        mapper.setSerializationInclusion(JsonInclude.Include.NON_NULL)
        mapper.configure(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS, true)
        mapper.configure(SerializationFeature.WRITE_ENUMS_USING_INDEX, false)
        mapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
        mapper.configure(JsonParser.Feature.ALLOW_SINGLE_QUOTES, true)
        mapper.configure(MapperFeature.USE_GETTERS_AS_SETTERS, false)
        mapper.configure(MapperFeature.ACCEPT_CASE_INSENSITIVE_ENUMS, true)
        mapper.dateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSXXX")
    }

    val mediaType: MediaType? = MediaType.parse("application/json; charset=utf-8")
}
